clippy_lints/src/implicit_transmute_types.rs

```diff
@@ -0,0 +1,108 @@
+    /// let transmuted = unsafe { core::mem::transmute(wrapped) };
+    ///
+    /// // This is sound now, but if it gets changed in the future to
+    /// // something that expects a type other than `char`, the transmute
+    /// // would infer it returns that type, which is likely unsound.
+    /// let _ = char::is_lowercase(transmuted);
+    /// ```
+    ///
+    /// Specify type parameters:
+    /// ```rust
+    /// # #[repr(transparent)]
+    /// # struct CharWrapper {
+    /// #     _inner: char,
+    /// # }
+    /// let wrapped = CharWrapper { _inner: 'a' };
+    /// // Because we explicitly specify the types for the transmute, any change in
+    /// // surrounding code that would cause the transmute call to infer different
+    /// // types will now be caught by typechecking, forcing us to come back and
+    /// // reassess the soundsness of transmuting between the new types.
+    /// let transmuted = unsafe { core::mem::transmute::<CharWrapper, char>(wrapped) };
+    ///
+    /// let _ = char::is_lowercase(transmuted);
+    /// ```
+    ///
+    /// If you decide that you *do* want the types to be inferred,
+    /// you can silence the lint by conveying your intention explicitly:
+    /// ```rust
+    /// # use std::mem::transmute;
+    /// # fn main() {
+    /// #     unsafe {
+    /// #         let foo: i32 = 123;
+    /// #         let _: u32 =
+    /// transmute::<_, _>(foo);
+    /// #     }
+    /// # }
+    /// ```
+    ///
+    /// [`transmute`]: https://doc.rust-lang.org/core/mem/fn.transmute.html
+    #[clippy::version = "1.72.0"]
+    pub IMPLICIT_TRANSMUTE_TYPES,
+    style,
+    "calling mem::transmute without explicit type parameters"
+}
+
+declare_lint_pass!(ImplicitTransmuteTypes => [IMPLICIT_TRANSMUTE_TYPES]);
+
+impl<'tcx> LateLintPass<'tcx> for ImplicitTransmuteTypes {
+    fn check_expr(&mut self, cx: &LateContext<'_>, e: &hir::Expr<'_>) {
+        if let ExprKind::Call(func, _) = &e.kind
+            && let ExprKind::Path(qpath) = &func.kind
+            && let Some(def_id) = cx.qpath_res(qpath, func.hir_id).opt_def_id()
+            && is_diagnostic_item_or_ctor(cx, def_id, sym::transmute)
+            && last_path_segment(qpath).args.is_none()
+        {
+            let suggestion_span = qpath.span().shrink_to_hi();
+
+            let substs = cx.typeck_results().node_substs(func.hir_id);
+            let srctype = substs.type_at(0);
+            let dsttype = substs.type_at(1);
+
+
+            span_lint_and_then(
+                cx,
+                IMPLICIT_TRANSMUTE_TYPES,
+                e.span,
+                "`transmute` called without explicit type parameters",
+                |b| {
+                    b.span_suggestion_verbose(
+                        suggestion_span,
+                        "consider specifying the types intended to be transmuted",
+                        format!("::<{srctype}, {dsttype}>"),
+                        Applicability::MachineApplicable
```

Comment:
Printing types like this can sometimes cause problems, when rustc can't determine the correct path for an item. This should be:
```suggestion
                        Applicability::MaybeIncorrect
```